sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ComplexTypes.scala

```diff
@@ -39,7 +41,18 @@ object SimplifyExtractValueOps extends Rule[LogicalPlan] {
       // Remove redundant field extraction.
       case GetStructField(createNamedStruct: CreateNamedStruct, ordinal, _) =>
         createNamedStruct.valExprs(ordinal)
-
+      case GetStructField(WithFields(struct, nameExprs, valExprs), ordinal, maybeName) =>
+        val extractFieldName = maybeName.getOrElse(
+          struct.dataType.asInstanceOf[StructType](ordinal).name)
+        val resolver = SQLConf.get.resolver
+        val names = nameExprs.map(e => e.eval().toString)
+        if (names.exists(n => resolver(n, extractFieldName))) {
```

Comment:
to be more conservative:
```
val matches = names.zip(valExprs).filter { case (name, _) => resolver(name, extractFieldName) }
if (matches.length == 1) {
  matches.head._2
} else {
  GetStructField(struct, ordinal, Some(extractFieldName))
}
```

Reply:
You're right, this is more concise. To give the same behaviour though, I had to change it slightly: 
```
val matches = names.zip(valExprs).filter { case (name, _) => resolver(name, extractFieldName) }
if (matches.nonEmpty) {
  matches.last._2
} else {
  GetStructField(struct, ordinal, Some(extractFieldName))
}
```
Reason: It's possible for there to be multiple matches and we only want the last one.